In [1]:
# Các gói cần dùng
import scipy.stats as st
import statsmodels
import statsmodels.api
import numpy as np
import pandas as pd


### 1. Ước lượng khoảng cho trung bình tổng thể

#### Ví Dụ 1:

Để ước tính trọng lượng của một con chuột, người ta chọn ngẫu nhiên 10 chú chuột. Dưới đây lần lượt là trọng lượng (gam) của 10 chú chuột trong mẫu.

21, 23, 27, 19, 17, 18, 20, 15, 17, 22

Giả sử trọng lượng của một con chuột tuân theo phân phối chuẩn. Từ
bộ dữ liệu này,

a. Đưa ra khoảng tin cậy 80% cho trọng lượng trung bình của các con chuột.

b. Đưa ra khoảng tin cậy 95% cho trọng lượng trung bình của các con chuột. So
sánh với kết quả tìm được trong câu (a) và đưa ra nhận xét.

In [2]:
# Lệnh dùng ULK cho TB tong the
?st.t.interval

In [4]:
# Câu a
mau=np.array([21, 23, 27, 19, 17, 18, 20, 15, 17, 22]) # nhập dữ liệu
st.t.interval(confidence=0.8, loc=np.mean(mau),
             scale=st.sem(mau), df=len(mau)-1)
# confidence: độ tin cậy của ước lượng
# loc: trung bình mẫu
# df: bậc tự do (= n-1 = cỡ mẫu - 1)

(18.364761959544772, 21.435238040455225)

Khoảng tin cậy 80% cho trọng lượng trung bình của các con chuột: [18.36, 21.44]

In [7]:
# Cách 2 tìm khoảng ước lượng cho trung bình tổng thể:
kiemdinh = st.ttest_1samp(mau, popmean=0) #Lệnh thực hiện bài toán kiểm định (sẽ học chi tiết sau)
kiemdinh.confidence_interval(confidence_level=0.8) #dùng method confidence_interval
# để tìm ước lượng khoảng cho trung bình từ kết quả của bài toán kiểm định
# Kết quả (18.36476195954477, 21.43523804045523) hoàn toàn trùng khớp với kq bên trên

ConfidenceInterval(low=18.36476195954477, high=21.43523804045523)

In [6]:
# Câu b.
st.t.interval(confidence=0.95,loc=np.mean(mau),scale=st.sem(mau),df=len(mau)-1)

(17.388881009258824, 22.411118990741173)

In [8]:
#Hoặc
kiemdinh.confidence_interval(confidence_level=0.95)

ConfidenceInterval(low=17.388881009258824, high=22.411118990741173)

Khoảng tin cậy 95% cho trọng lượng trung bình của các con chuột: [17.39, 22.41]

Nhận xét: Độ tin cậy càng lớn thì khoảng ước lượng càng rộng.

#### Ví dụ 2:

Từ dữ liệu ThoiGianCho.csv hãy tìm khoảng tin cậy 90% cho thời gian chờ trung bình của tất cả các lượt khách hàng.

In [10]:
tgc=pd.read_csv('ThoiGianCho.csv')

In [20]:
print(tgc.head())

   ThoiGian
0       1.6
1       6.6
2       5.6
3       5.1
4       3.9


In [9]:
st.t.interval(confidence=0.9,loc=np.mean(tgc.ThoiGian),
              scale=st.sem(tgc.ThoiGian),df=len(tgc.ThoiGian)-1)

(5.048977596460333, 5.871022403539665)

In [12]:
#Hoặc cách khác
btkd = st.ttest_1samp(tgc.ThoiGian,popmean=0)
btkd.confidence_interval(confidence_level=0.9)

ConfidenceInterval(low=5.048977596460333, high=5.871022403539665)

Khoảng ước lượng cần tìm là [5.05, 5.87]

### 2. Khoảng tin cậy cho tỉ lệ

In [2]:
#Lệnh tìm khoảng tin cậy cho tỉ lệ
?statsmodels.stats.proportion.proportion_confint # xem chi tiết
statsmodels.stats.proportion.proportion_confint # xem tham số

<function statsmodels.stats.proportion.proportion_confint(count, nobs, alpha=0.05, method='normal')>

#### Ví dụ 1:

Chọn ngẫu nhiên 1200 trẻ sơ sinh ở Việt Nam trong năm 2020 có 625 bé trai, 575
bé gái 1 và 120 bé là con thứ ba trong gia đình. 

a. Tìm ước lượng điểm cho tỷ lệ sinh con trai, tỷ lệ sinh con gái và tỷ lệ sinh con thứ ba của Việt Nam năm 2020.

Các ước lượng điểm lần lượt là 625/1200; 575/1200 và 120/1200

b. Tìm khoảng ước lượng 99% cho tỷ lệ sinh con trai của Việt Nam năm 2020.

In [14]:
statsmodels.stats.proportion.proportion_confint(count=625,
                   nobs = 1200, alpha=0.01)
# count: số quan sát thành công (trong trường hợp này là số con trai)
# nobs: cỡ mẫu
# alpha = 1 - độ tin cậy (trong trường hợp này là 1 - 0.99 = 0.01)

(0.48368672715336297, 0.5579799395133037)

Kết quả: [0.48, 0.56]

#### Ví dụ 2:

Từ tập dữ liệu mẫu SoLieu.csv hãy tìm khoảng tin cậy 90% cho tỉ lệ dân số ở khu vực thành phố.

In [16]:
from statsmodels.stats.proportion import proportion_confint
# Làm cho câu lệnh ước lượng khoảng tỷ lệ đỡ dài

In [21]:
SL=pd.read_csv('SoLieu.csv')
SL.head()
SL.KhuVuc.value_counts()

ThanhPho    42
MienNui     26
NongThon    20
HaiDao      12
Name: KhuVuc, dtype: int64

In [22]:
proportion_confint(count=42, 
                   nobs=len(SL.KhuVuc),alpha=0.1)

(0.3388168499368469, 0.5011831500631531)

#### Ví dụ 3:

Tìm khoảng tin cậy 95% cho ti le người dân có thu nhập trên 75 (dùng tập dữ liệu mẫu SoLieu.csv).

In [23]:
len(SL.ThuNhap) # tìm nobs (cỡ mẫu)

100

In [25]:
np.sum(SL.ThuNhap>75) # tìm count (số quan sát thành công)
# hoặc dùng lệnh len(SL.ThuNhap[SL.ThuNhap>75])

15

In [26]:
proportion_confint(count=15,nobs=100,alpha=0.05)

(0.08001528740942766, 0.21998471259057234)